crates/autopilot/src/decoded_settlement.rs

```diff
@@ -175,16 +175,21 @@ impl DecodedSettlement {
         })
     }
 
-    // Assumes it is called with already FILLED orders.
-    // Needs rework to support partially fillable orders.
-    // Tricky because the decoded settlement is using FILLED `orders` so we don't
-    // always know the executed amount in case of partial fill.
-    pub fn total_fees(&self, external_prices: &ExternalPrices, orders: &[Order]) -> U256 {
+    /// Returns the total `executed_solver_fee` of this solution converted to
+    /// the native token. This is only the value used for objective value
+    /// computatations and can theoretically be different from the value of
+    /// fees actually collected by the protocol.
+    pub fn total_fees(&self, external_prices: &ExternalPrices, orders: &[OrderExecution]) -> U256 {
         self.trades.iter().fold(0.into(), |acc, trade| {
-            match orders
-                .iter()
-                .find(|order| order.signature == trade.signature.0)
-            {
+            match orders.iter().find(|order| {
+                // I think theoretically it's possible that the same partiallye fillable order
+                // gets filled twice for the same `executed_amount` in the same
+                // auction with different `executed_solver_fee`s but I'm not
+                // sure this can be made unambiguous here and I'm not sure if it
+                // would practically ever happen.
+                order.signature == trade.signature.0
+                    && order.executed_amount == trade.executed_amount
```

Comment:
Is there supposed to be a one to one correspondence of traders and order-executions? I see that below we're warning if an order wasn't found. Should that be an error?
If there is a one to one correspondence then we could remove a used order-executions so that it can't be matched again. The fee is always added so it doesn't matter which one you match first as long as you eventually  match both.

Reply:
> Is there supposed to be a one to one correspondence of traders and order-executions?

Ideally yes. If this is not the case our DB is missing some data AFAIK.

> I see that below we're warning if an order wasn't found. Should that be an error?

Probably no since it's likely more helpful to have some incomplete data instead of no data due to erroring out.

> If there is a one to one correspondence then we could remove a used order-executions so that it can't be matched again. The fee is always added so it doesn't matter which one you match first as long as you eventually match both.

Sounds good, will do. This might be incorrect in case we have for some reason more trades in our DB than on-chain but l think we probably don't have to worry about that.